math/tangent.go

```diff
@@ -1,132 +1,114 @@
-EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
-OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT
-SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
-SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
-SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
-HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
-OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
-SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
-*/
 import (
-	stdMath "math"
-
 	"github.com/ericlagergren/decimal"
+	"github.com/ericlagergren/decimal/internal/arith"
+	"github.com/ericlagergren/decimal/internal/arith/checked"
+	"github.com/ericlagergren/decimal/misc"
 )
 
-func prepareTangentInput(precision int, theta *decimal.Big) (*decimal.Big, bool) {
-	cPi := Pi(decimal.WithPrecision(precision))
-	cPiOver2 := decimal.WithPrecision(precision).Quo(cPi, two)
-	var x *decimal.Big
-	if cPiOver2.CmpAbs(theta) < 0 {
-		//for cos to work correctly the input
-		// must be |theta|< 2Pi so we'll fix it
-		var m *decimal.Big
-		if theta.Signbit() {
-			m = decimal.WithPrecision(precision).Add(theta, cPiOver2)
+func prepTan(z, x *decimal.Big, ctx decimal.Context) (*decimal.Big, bool) {
+	x0 := alias(z, x)
+	if pi2 := pi2(x0, ctx); x.CmpAbs(pi2) >= 0 {
+		// for cos to work correctly the input must be in (-2pi, 2pi)
+		if x.Signbit() {
+			ctx.Add(x0, x, pi2)
 		} else {
-			m = decimal.WithPrecision(precision).Sub(theta, cPiOver2)
+			ctx.Sub(x0, x, pi2)
 		}
-		m = decimal.WithPrecision(precision).QuoInt(m, cPi)
-		if theta.Signbit() {
-			m.Sub(m, one)
+
+		var tmp decimal.Big
+		ctx.QuoInt(x0, x0, Pi(&tmp, ctx))
+		if x.Signbit() {
+			ctx.Sub(x0, x0, one)
 		} else {
-			m.Add(m, one)
+			ctx.Add(x0, x0, one)
 		}
-		mInt, ok := m.Int64()
+
+		v, ok := x0.Int64()
 		if !ok {
 			return nil, false
 		}
-		//now we'll resize Pi to be a more accurate precision
-		piPrecision := precision + int(stdMath.Ceil(stdMath.Abs(float64(mInt))/float64(10)))
-		toRemove := Pi(decimal.WithPrecision(piPrecision))
-		toRemove = toRemove.Mul(toRemove, m)
+		uv := arith.Abs(v)
 
-		//so toRemove = mPi so
-		// |theta-toRemove| < Pi/2
-		//add 1 to the precision for the up comming squaring
-		x = decimal.WithPrecision(precision+1).Sub(theta, toRemove)
+		// Adjust so we have ceil(v/10) + ctx.Precision, but check for overflows.
+		// 1+((v-1)/10) will be widly incorrect for v == 0, but x/y = 0 iff
+		// x = 0 and y != 0. In this case, -2pi <= x >= 2pi, so we're fine.
+		prec, ok := checked.Add(1+((uv-1)/10), uint64(ctx.Precision))
+		if !ok || prec > maxInt {
+			return nil, false
+		}
+		pctx := decimal.Context{Precision: int(prec)}
+
+		pctx.Mul(x0, Pi(&tmp, pctx), x0)
+
+		ctx.Precision++
```

Comment:
This increase of precision `ctx.Precision++` can be removed.  it came from the `precision+1` in my code but this was something that i meant to remove and forgot.  Doesn't hurt accuracy to have it but it more to compute.

Reply:
Do I need to keep the `ctx.Precision++` in `cosine.go`? Or can it be dropped as well?

Reply:
That one is correct, so leave it.  The line 56 `ctx.Precision++` in cosine.go is needed because of the squaring that occurs one line 94.  There might be the case if halfed >1 that it isn't needed but I'd have to think about that for a bit and work out the math bits to see where that line is.  For now I would say leave it.